OpenStreetMap Sample Update - Removed Jupyter Dashboard

Jwmazzi opened the pull request:
I've updated the OpenStreetMap sample to remove the reliance on Jupyter Dashboard.
-----

# Checklist

Please go through each entry in the below checklist and mark an 'X' if that condition has been met. Every entry should be marked with an 'X' to be get the Pull Request approved.


- [X] All `import`s are in the first cell? First block of imports are standard libraries, second block are 3rd party libraries, third block are all `arcgis` imports?
- [X] All `GIS` object instantiations are one of the following?
    - `gis = GIS('https://www.arcgis.com', 'arcgis_python', 'P@ssword123')`
    - `gis = GIS(profile="your_online_profile")`
    - `gis = GIS('https://pythonapi.playground.esri.com/portal', 'arcgis_python', 'amazing_arcgis_123')`
    - `gis = GIS(profile="your_enterprise_portal")`
- [X] If this notebook requires setup or teardown, did you add the appropriate code to `./misc/setup.py` and/or `./misc/teardown.py`?
- [X] If this notebook references any portal items that need to be staged on AGOL/Python API playground, did you coordinate with a Python API team member to stage the item the correct way with the api\_data\_owner user?
- [X] Code refactored & split out across multiple cells, useful comments?
- [X] Consistent voice/tense/narrative style? Thoroughly checked for typos?
- [X] All images used like `<img src="base64str_here">` instead of `<img src="https://some.url">`? All map widgets contain a static image preview? (Call `mapview_inst.take_screenshot()` to do so)
- [X] All file paths are constructed in an OS-agnostic fashion with `os.path.join()`? (Instead of `r"\foo\bar"`, `os.path.join(os.path.sep, "foo", "bar")`, etc.)
- [X] **IF YOU WANT THIS SAMPLE TO BE DISPLAYED ON THE DEVELOPERS.ARCGIS.COM WEBSITE**, ping @ DavidJVitale so he can add it to the list for the next deploy

AtmaMani commented:
@jyaistMap could you review and merge this PR

Jwmazzi commented:
Greetings. This sample might be useful for the upcoming FedGIS conference. Is there anything I can do to get it past the final stages of this review?

jyaistMap commented:
@Jwmazzi I'll finish up the review today and get it to the next stage.

jyaistMap commented on samples/02_power_users_developers/openstreetmap_exploration.ipynb:
```diff
@@ -0,0 +1,1338 @@
+{
```
<p>I reviewed this notebook by running it locally in 2 different environments:</p><ul><li> one with <em>arcpy</em> created by cloning the default ArcGIS Pro 2.4.2 environment and then upgrading the <em>arcgis</em> to 1.7 and installing <em>osm_runner</em></li><li>one without <em>arcpy</em> by installing <em>arcgis</em> from the latest source code in the git repo and installing <em>osm_runner</em></li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/551/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

jyaistMap replied:
<p>I think it might clarify to use one cell to display the data and a separate cell to assign the variables. Creating the <em>full_extent </em>variable and then using <em>project() </em>on it, I expected the output to be in lat/long values because of the <em>out_sr=4326 </em>parameter.  I  read through the code again to discover that I wasn't actually displaying the values for any of those variables I created after projecting but instead was displaying the spatial dataframe that I had downloaded.  It might be helpful to explicitly state that the <em>gen_osm_df() </em>method expects a parameter in a specific format to explain why you are creating these variables.</p><p> </p><ul><li>NOTE: The narrative guides me that I could use <em>dc_df_spatial_project() </em>if I have an ArcPy enviroment. When I use this function, the output in the SHAPE column actually displays the values in lat/long coordinates, not Web Mercator coordinates.  It could be helpful to clarify whether the dataframe needs to be projected or not depending upon the environment</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/551/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

jyaistMap replied:
<ul><li>in the environment without <em>ArcPy, </em>when I first ran this cell I receive a value of -685 Square Kilometers</li><li>I installed the <em>shapely </em>package and that corrected it initially, but now currently I receive negative values for the area queries above and the population density calculations below.</li><li>from research I've done this is usually because of ring ordering in the polygons, but I'm not sure what to add  to my environment or how to correct it</li></ul><p> </p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/551/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

jyaistMap replied:
<ul><li>in the environment without <em>ArcPy, </em>I receive a value of -173 Square Kilometers</li><li>I installed the <em>shapely </em>package and that corrected it initially, but now I'm seeing this value after running it again.</li><li>I'm assuming with the correct modules this doesn't occur - did you notice negative values at any time?</li></ul><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/551/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

jyaistMap replied:
<ul><li>I suggested hyperlinking to the <a href="https://developers.arcgis.com/python/api-reference/arcgis.geoenrichment.html#enrich" target="_blank"><em>enrich()</em></a> api function for those users unfamiliar with it</li><li>I think it would help to link to the<a href="https://developers.arcgis.com/rest/geoenrichment/api-reference/geoenrichment-service-overview.htm" target="_blank"> Geoenrichment Service</a> and the <a href="https://developers.arcgis.com/rest/geoenrichment/api-reference/enrich.htm" target="_blank">Enrich </a>task so users can investigate what attributes get added to the data when no variables are entered into the <a href="https://developers.arcgis.com/python/api-reference/arcgis.geoenrichment.html#enrich" target="_blank"><em>enrich()</em></a> function.</li><li>change <em>assesed </em>to <em>assessed</em></li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/551/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

jyaistMap replied:
<p>Change last sentence to something like:</p><p><em>We are adding the following results to the first 2 maps we created above.</em></p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/551/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

jyaistMap replied:
<p>Change <em>Eurproe</em> to <em>Europe</em></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/551/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

jyaistMap replied:
<p>Seems the word cameras might be a typo in this sentence:   &nbsp;<em>Locate a place, drag the map around, and then run the last cell to plot a heat map of recycling amenities cameras and all of the shops within the extent of the map.</em>&nbsp;</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/551/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

jyaistMap replied:
<p>change <em>conveinent </em>to <em>convenient</em></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/551/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='19'/>

jyaistMap replied:
<ul><li>In the environment with <em>arcpy</em>,  all cells run successfully, except the last one:</li><li>I continually receive the following error when running the <em>to_featurelayer() </em>command</li></ul><p> </p><p><strong>---------------------------------------------------------------------------</strong></p><p><strong>RuntimeError</strong>                              Traceback (most recent call last)</p><p><strong>&lt;ipython-input-28-e9241c877088&gt;</strong> in &lt;module&gt;</p><p><strong>----&gt; 1 </strong>recycling_hfl <strong>=</strong> recy_df<strong>.</strong>spatial<strong>.</strong>to_featurelayer<strong>(f'OSM_Recycling_{round(time.time())}',</strong> gis<strong>=</strong>gis<strong>,</strong> tags<strong>='OSM')</strong></p><p>      2 shops_hfl <strong>=</strong> shop_df<strong>.</strong>spatial<strong>.</strong>to_featurelayer<strong>(f'OSM_Shops_{round(time.time())}',</strong> gis<strong>=</strong>gis<strong>,</strong> tags<strong>='OSM')</strong></p><p>      3 </p><p>      4 display<strong>(</strong>recycling_hfl<strong>)</strong></p><p>      5 display<strong>(</strong>shops_hfl<strong>)</strong></p><p> </p><p><strong>~\AppData\Local\ESRI\conda\envs\arcgispro-py3-clone\lib\site-packages\arcgis\features\geo\_accessor.py</strong> in to_featurelayer<strong>(self, title, gis, tags, folder)</strong></p><p>   1983                 <strong>raise</strong> ValueError<strong>("GIS object must be provided")</strong></p><p>   1984         content <strong>=</strong> gis<strong>.</strong>content</p><p><strong>-&gt; 1985         return</strong> content<strong>.</strong>import_data<strong>(</strong>self<strong>.</strong>_data<strong>,</strong> folder<strong>=</strong>folder<strong>,</strong> title<strong>=</strong>title<strong>,</strong> tags<strong>=</strong>tags<strong>)</strong></p><p>   1986     <strong># ----------------------------------------------------------------------</strong></p><p>   1987     <strong>@</strong>staticmethod</p><p> </p><p><strong>~\AppData\Local\ESRI\conda\envs\arcgispro-py3-clone\lib\site-packages\arcgis\gis\__init__.py</strong> in import_data<strong>(self, df, address_fields, folder, **kwargs)</strong></p><p>   4114                     ds <strong>=</strong>\</p><p>   4115                         df.spatial.to_featureclass(location=os.path.join(fgdb,</p><p><strong>-&gt; 4116                                                                          os.path.basename(temp_dir)))</strong></p><p>   4117 </p><p>   4118                 zip_fgdb <strong>=</strong> zipws<strong>(</strong>path<strong>=</strong>fgdb<strong>,</strong> outfile<strong>=</strong>temp_zip<strong>,</strong> keep<strong>=True)</strong></p><p> </p><p><strong>~\AppData\Local\ESRI\conda\envs\arcgispro-py3-clone\lib\site-packages\arcgis\features\geo\_accessor.py</strong> in to_featureclass<strong>(self, location, overwrite)</strong></p><p>   1923         return to_featureclass(self,</p><p>   1924                                location<strong>=</strong>location<strong>,</strong></p><p><strong>-&gt; 1925                                overwrite=overwrite)</strong></p><p>   1926     <strong>#----------------------------------------------------------------------</strong></p><p>   1927     <strong>def</strong> to_table<strong>(</strong>self<strong>,</strong> location<strong>,</strong> overwrite<strong>=True):</strong></p><p> </p><p><strong>~\AppData\Local\ESRI\conda\envs\arcgispro-py3-clone\lib\site-packages\arcgis\features\geo\_io\fileops.py</strong> in to_featureclass<strong>(geo, location, overwrite, validate)</strong></p><p>    610 </p><p>    611         array <strong>=</strong> np<strong>.</strong>array<strong>([],</strong> np<strong>.</strong>dtype<strong>(</strong>dtypes<strong>))</strong></p><p><strong>--&gt; 612         </strong>arcpy<strong>.</strong>da<strong>.</strong>ExtendTable<strong>(</strong>fc<strong>,</strong> oidfld<strong>,</strong> array<strong>,</strong> join_dummy<strong>,</strong> append_only<strong>=False)</strong></p><p>    613 </p><p>    614         <strong># 3. Insert the Data</strong></p><p> </p><p><strong>RuntimeError</strong>: The operation was attempted on an empty geometry.</p><p> </p><ul><li>I tried altering the geometry column by running <em>df.spatial.set_geometry(col='SHAPE')</em>, but I still received the error</li><li>I also received an error (but have been unable to reproduce) about a Field name being invalid, which I assumed resulted because of the colons in certain field names in the <em>shop_df</em> and <em>recy_df</em> frames. I replaced the colons with underscores and still received the above conda error.</li></ul><p> </p><ul><li>in the environment without arcpy, the feature layers are created correctly</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/551/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='20'/>

jyaistMap commented:
@Jwmazzi @AtmaMani @Yongyao 
* I've run this notebook a few more times this morning in 3 different environments:
  * MacOS conda environment:
    - Initially, I created the environment from the `envrionment.yml` file in the source repo, then used `pip` to install `arcgis` and `osm-runner`. All initial area calculations returned as negative. I installed `shapely` into the environment and the area calculations returned correctly. The notebook ran perfectly.
  * Windows conda environment (without `arcpy`):
     * Created  the environment in same manner as above. Same experience.
  * Windows conda enviroment (with `arcpy`):
    * I cloned the default `arcgispro-py3` environment installed with `ArcGIS Pro 2.4.2 `, then used pip to install `osm-runner`. The notebook runs except for the last cell. Iniitally I get the following error:
```pytb
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-11-e9241c877088> in <module>
----> 1 recycling_hfl = recy_df.spatial.to_featurelayer(f'OSM_Recycling_{round(time.time())}', gis=gis, tags='OSM')
      2 shops_hfl = shop_df.spatial.to_featurelayer(f'OSM_Shops_{round(time.time())}', gis=gis, tags='OSM')
      3 
      4 display(recycling_hfl)
      5 display(shops_hfl)

~\AppData\Local\ESRI\conda\envs\arcgispro-py3-clone\lib\site-packages\arcgis\features\geo\_accessor.py in to_featurelayer(self, title, gis, tags, folder)
   1983                 raise ValueError("GIS object must be provided")
   1984         content = gis.content
-> 1985         return content.import_data(self._data, folder=folder, title=title, tags=tags)
   1986     # ----------------------------------------------------------------------
   1987     @staticmethod

~\AppData\Local\ESRI\conda\envs\arcgispro-py3-clone\lib\site-packages\arcgis\gis\__init__.py in import_data(self, df, address_fields, folder, **kwargs)
   4114                     ds =\
   4115                         df.spatial.to_featureclass(location=os.path.join(fgdb,
-> 4116                                                                          os.path.basename(temp_dir)))
   4117 
   4118                 zip_fgdb = zipws(path=fgdb, outfile=temp_zip, keep=True)

~\AppData\Local\ESRI\conda\envs\arcgispro-py3-clone\lib\site-packages\arcgis\features\geo\_accessor.py in to_featureclass(self, location, overwrite)
   1923         return to_featureclass(self,
   1924                                location=location,
-> 1925                                overwrite=overwrite)
   1926     #----------------------------------------------------------------------
   1927     def to_table(self, location, overwrite=True):

~\AppData\Local\ESRI\conda\envs\arcgispro-py3-clone\lib\site-packages\arcgis\features\geo\_io\fileops.py in to_featureclass(geo, location, overwrite, validate)
    610 
    611         array = np.array([], np.dtype(dtypes))
--> 612         arcpy.da.ExtendTable(fc, oidfld, array, join_dummy, append_only=False)
    613 
    614         # 3. Insert the Data

RuntimeError: The name of the Field is invalid: valid names may contain letters, numbers or underscores.
```

* so
  * I concluded the colons in the field names resulting from `OSM` caused the problem so then I renamed the columns replacing the colons with underscores, 
which then I received:
```pytb
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-48-e9f079377d14> in <module>
----> 1 recycling_hfl = recy_df.spatial.to_featurelayer(f'OSM_Recycling_{round(time.time())}', gis=gis, tags='OSM')

~\AppData\Local\ESRI\conda\envs\arcgispro-py3-clone\lib\site-packages\arcgis\features\geo\_accessor.py in to_featurelayer(self, title, gis, tags, folder)
   1983                 raise ValueError("GIS object must be provided")
   1984         content = gis.content
-> 1985         return content.import_data(self._data, folder=folder, title=title, tags=tags)
   1986     # ----------------------------------------------------------------------
   1987     @staticmethod

~\AppData\Local\ESRI\conda\envs\arcgispro-py3-clone\lib\site-packages\arcgis\gis\__init__.py in import_data(self, df, address_fields, folder, **kwargs)
   4114                     ds =\
   4115                         df.spatial.to_featureclass(location=os.path.join(fgdb,
-> 4116                                                                          os.path.basename(temp_dir)))
   4117 
   4118                 zip_fgdb = zipws(path=fgdb, outfile=temp_zip, keep=True)

~\AppData\Local\ESRI\conda\envs\arcgispro-py3-clone\lib\site-packages\arcgis\features\geo\_accessor.py in to_featureclass(self, location, overwrite)
   1923         return to_featureclass(self,
   1924                                location=location,
-> 1925                                overwrite=overwrite)
   1926     #----------------------------------------------------------------------
   1927     def to_table(self, location, overwrite=True):

~\AppData\Local\ESRI\conda\envs\arcgispro-py3-clone\lib\site-packages\arcgis\features\geo\_io\fileops.py in to_featureclass(geo, location, overwrite, validate)
    610 
    611         array = np.array([], np.dtype(dtypes))
--> 612         arcpy.da.ExtendTable(fc, oidfld, array, join_dummy, append_only=False)
    613 
    614         # 3. Insert the Data

RuntimeError: The operation was attempted on an empty geometry.
```

* Any environment with `arcpy` that I've tried running the notebook in will not execute the last cell. Whether or not receiving an error regarding invalid field names (I do not always get this message), the `RuntimeError: The operation was attempted on an empty geometry.` always occurs.
* Without `arcpy`, the notebook runs, and provided `shapely` is in the envrinment, all area calculations appear as positive values.

Jwmazzi commented:
I updated the typos and added a line at the beginning that tells the user this Notebook was written for an environment without arcpy. I'll go through the other things observed (differences in areas and so forth) when I get get a chance. 

After seeing so many discrepancies with arcpy being present or not, it is better to just scrap this example?

jyaistMap commented:
@Jwmazzi Adding the note that the notebook should only be fun in an environment without `arcpy` should suffice.  It might also help to remove the sentence _If our environment had access to ArcPy, we could have simply used the project() method of the GeoAccessor; e.g. dc_df.spatial.project()._ at the end of the **Build DataFrames....** section, too to not lead users down that path.

Jwmazzi commented:
Removed that line about the project method.

AtmaMani commented:
@jyaistMap the sample looks good to me. Could you review if the latest changes addresses your suggestion and if so, merge it?
